From b879333c6f074dab1fc3dad880cadfd7d2d70012 Mon Sep 17 00:00:00 2001
From: Sudharsan Eswaran <eswaran@qti.qualcomm.com>
Date: Tue, 2 Jun 2026 15:13:01 +0530
Subject: [PATCH] AWGN event handling support

1. support to get the AWGN event in NL & whm HAL
2. process the event and forward it to channel selection task
3. Mark the preference for AWGN detected channel to 0 and set reason to 14

Signed-off-by: Sudharsan Eswaran <eswaran@qti.qualcomm.com>
---
 .../ap_manager/ap_manager.cpp                 | 27 +++++++
 agent/src/beerocks/slave/agent_db.h           |  1 +
 agent/src/beerocks/slave/son_slave_thread.cpp | 30 +++++++
 .../slave/tasks/channel_selection_task.cpp    | 80 ++++++++++++++++++-
 .../slave/tasks/channel_selection_task.h      |  9 +++
 common/beerocks/bwl/include/bwl/ap_wlan_hal.h |  1 +
 .../bwl/include/bwl/base_wlan_hal_types.h     |  9 +++
 .../bwl/nl80211/ap_wlan_hal_nl80211.cpp       | 31 +++++++
 common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp   | 33 ++++++++
 .../tlvf/beerocks_message_action.yaml         |  6 +-
 .../tlvf/beerocks_message_apmanager.yaml      |  5 ++
 .../tlvf/beerocks_message_backhaul.yaml       |  5 ++
 .../tlvf/structs/sAwgnDetectedParams.yaml     |  8 ++
 .../tlvf/wfa_map/tlvChannelPreference.yaml    |  1 +
 14 files changed, 243 insertions(+), 3 deletions(-)
 create mode 100644 common/beerocks/tlvf/yaml/beerocks/tlvf/structs/sAwgnDetectedParams.yaml

diff --git a/agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager.cpp b/agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager.cpp
index 84b08f73f..43347ae6e 100644
--- a/agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager.cpp
+++ b/agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager.cpp
@@ -2780,6 +2780,33 @@ bool ApManager::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t event_ptr)
         }
     } break;
 
+    case Event::AWGN_Detected: {
+        if (!data) {
+            LOG(ERROR) << "AWGN_Detected without data!";
+            return false;
+        }
+
+        auto msg = static_cast<bwl::sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION *>(data);
+        LOG(INFO) << "AWGN detected on channel " << int(msg->params.channel)
+                  << ", Freq " << int(msg->params.frequency);
+
+        // Forward to slave thread (son_slave_thread)
+        auto notification = message_com::create_vs_message<
+            beerocks_message::cACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION>(cmdu_tx);
+
+        if (!notification) {
+            LOG(ERROR) << "Failed to create AWGN notification message";
+            return false;
+        }
+
+        notification->params().channel = msg->params.channel;
+        notification->params().frequency = msg->params.frequency;
+
+        LOG(INFO) << "AWGN notified to slave thread";
+        // Send to slave
+        send_cmdu(cmdu_tx);
+    } break;
+
     // STA Connected
     case Event::STA_Connected: {
 
diff --git a/agent/src/beerocks/slave/agent_db.h b/agent/src/beerocks/slave/agent_db.h
index ceb2c3800..24c590c20 100644
--- a/agent/src/beerocks/slave/agent_db.h
+++ b/agent/src/beerocks/slave/agent_db.h
@@ -350,6 +350,7 @@ public:
             int8_t tx_power_dbm;
             std::vector<beerocks_message::sSupportedBandwidth> supported_bw_list;
             beerocks_message::eDfsState dfs_state;
+            bool awgn_detected = false;
         };
 
         son::wireless_utils::sSpatialReuseParams spatial_reuse_params;
diff --git a/agent/src/beerocks/slave/son_slave_thread.cpp b/agent/src/beerocks/slave/son_slave_thread.cpp
index c67879dac..9f7de8e21 100644
--- a/agent/src/beerocks/slave/son_slave_thread.cpp
+++ b/agent/src/beerocks/slave/son_slave_thread.cpp
@@ -3477,6 +3477,36 @@ bool slave_thread::handle_cmdu_ap_manager_message(const std::string &fronthaul_i
         m_backhaul_manager_client->send_cmdu(cmdu_tx);
         break;
     }
+    case beerocks_message::ACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION: {
+        auto notification_in = beerocks_header->addClass<
+            beerocks_message::cACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION>();
+        if (!notification_in) {
+            LOG(ERROR) << "addClass ACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION failed";
+            return false;
+        }
+
+        LOG(TRACE) << "received ACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION";
+
+        auto notification_out = message_com::create_vs_message<
+            beerocks_message::cACTION_BACKHAUL_HOSTAP_AWGN_DETECTED_NOTIFICATION>(cmdu_tx);
+        if (!notification_out) {
+            LOG(ERROR) << "Failed building message!";
+            return false;
+        }
+        notification_out->params() = notification_in->params();
+
+        auto db    = AgentDB::get();
+        auto radio = db->radio(fronthaul_iface);
+        if (!radio) {
+            LOG(ERROR) << "AWGN: NULL radio";
+            break;
+        }
+        auto action_header         = message_com::get_beerocks_header(cmdu_tx)->actionhdr();
+        action_header->radio_mac() = radio->front.iface_mac;
+
+        m_backhaul_manager_client->send_cmdu(cmdu_tx);
+        break;
+    }
     case beerocks_message::ACTION_APMANAGER_HOSTAP_DFS_CAC_COMPLETED_NOTIFICATION: {
         auto notification_in = beerocks_header->addClass<
             beerocks_message::cACTION_APMANAGER_HOSTAP_DFS_CAC_COMPLETED_NOTIFICATION>();
diff --git a/agent/src/beerocks/slave/tasks/channel_selection_task.cpp b/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
index 87f9ba56d..cec9ef97a 100644
--- a/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
+++ b/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
@@ -618,6 +618,11 @@ bool ChannelSelectionTask::handle_vendor_specific(ieee1905_1::CmduMessageRx &cmd
         handle_vs_zwdfs_ant_channel_switch_response(cmdu_rx, sd, beerocks_header);
         break;
     }
+    case beerocks_message::ACTION_BACKHAUL_HOSTAP_AWGN_DETECTED_NOTIFICATION: {
+        LOG(INFO) << "ChannelSelectionTask: Handling AWGN notification";
+        handle_vs_awgn_detected_notification(cmdu_rx, sd, beerocks_header);
+        break;
+    }
 
     default: {
         // Message was not handled, therfore return false.
@@ -627,6 +632,67 @@ bool ChannelSelectionTask::handle_vendor_specific(ieee1905_1::CmduMessageRx &cmd
     return true;
 }
 
+void ChannelSelectionTask::handle_vs_awgn_detected_notification(
+    ieee1905_1::CmduMessageRx &cmdu_rx, int fd, std::shared_ptr<beerocks_header> beerocks_header)
+{
+    auto notification =
+        beerocks_header
+            ->addClass<beerocks_message::cACTION_BACKHAUL_HOSTAP_AWGN_DETECTED_NOTIFICATION>();
+    if (!notification) {
+        LOG(ERROR) << "addClass cACTION_BACKHAUL_HOSTAP_AWGN_DETECTED_NOTIFICATION failed";
+        return;
+    }
+
+    auto db = AgentDB::get();
+    auto radio =
+        db->get_radio_by_mac(beerocks_header->actionhdr()->radio_mac(), AgentDB::eMacType::RADIO);
+    if (!radio) {
+        return;
+    }
+
+    const auto &sender_iface_name = radio->front.iface_name;
+    const auto channel            = notification->params().channel;
+
+    LOG(INFO) << "Received AWGN_DETECTED notification from " << sender_iface_name
+              << " for channel " << int(channel);
+
+    // Mark channel as AWGN detected in the database
+    auto channel_it = radio->channels_list.find(channel);
+    if (channel_it != radio->channels_list.end()) {
+        channel_it->second.awgn_detected = true;
+        LOG(INFO) << "Marked channel " << int(channel) << " as AWGN detected in Agent DB";
+    } else {
+        LOG(WARNING) << "Channel " << int(channel) << " not found in channels list for radio "
+                     << sender_iface_name;
+    }
+
+    // Trigger Channel Preference Report
+    m_pending_preference.preference_ready.clear();
+    m_pending_preference.preference_ready[radio->front.iface_mac] = false;
+    m_send_preference_report_after_awgn_event = true;
+
+    auto agent_fd = m_btl_ctx.get_agent_fd();
+    if (agent_fd == beerocks::net::FileDescriptor::invalid_descriptor) {
+        LOG(ERROR) << "socket to Agent not found";
+        return;
+    }
+
+    LOG(DEBUG) << "Requesting updated channel list for AWGN event handling";
+
+    // Request updated channel list
+    auto request = message_com::create_vs_message<
+        beerocks_message::cACTION_BACKHAUL_CHANNELS_LIST_REQUEST>(m_cmdu_tx);
+    if (!request) {
+        LOG(ERROR) << "Failed to build message";
+        return;
+    }
+
+    auto action_header         = message_com::get_beerocks_header(m_cmdu_tx)->actionhdr();
+    action_header->radio_mac() = radio->front.iface_mac;
+
+    m_btl_ctx.send_cmdu(agent_fd, m_cmdu_tx);
+}
+
 void ChannelSelectionTask::handle_vs_csa_notification(
     ieee1905_1::CmduMessageRx &cmdu_rx, int fd, std::shared_ptr<beerocks_header> beerocks_header)
 {
@@ -983,7 +1049,8 @@ void ChannelSelectionTask::handle_vs_channels_list_response(
         ZWDFS_FSM_MOVE_STATE(eZwdfsState::CHOOSE_NEXT_BEST_CHANNEL);
     } else if (is_there_a_pending_preference || m_send_preference_report_after_cac_started_event ||
                m_send_preference_report_after_cac_completion_event ||
-               m_send_preference_report_after_csa_finished_event) {
+               m_send_preference_report_after_csa_finished_event ||
+               m_send_preference_report_after_awgn_event) {
 
         // If there is a pending preference query, need to build a preference report
         build_channel_preference_report(radio_mac);
@@ -1006,6 +1073,7 @@ void ChannelSelectionTask::handle_vs_channels_list_response(
             m_send_preference_report_after_cac_started_event    = false;
             m_send_preference_report_after_cac_completion_event = false;
             m_send_preference_report_after_csa_finished_event   = false;
+            m_send_preference_report_after_awgn_event           = false;
         }
     }
 }
@@ -1152,6 +1220,16 @@ bool ChannelSelectionTask::build_channel_preference_report(const sMacAddr &radio
                 wfa_map::cPreferenceOperatingClasses::eReasonCode::UNSPECIFIED);
         }
 
+        // Check for AWGN detection info
+        if (supported_channel_info.awgn_detected) {
+            LOG(DEBUG) << "Channel " << int(channel) << " in operating class " << int(operating_class)
+                       << " is NON_OPERABLE due to AWGN detection";
+            return AgentDB::sChannelPreference(
+                operating_class,
+                wfa_map::cPreferenceOperatingClasses::ePreference::NON_OPERABLE,
+                wfa_map::cPreferenceOperatingClasses::eReasonCode::AWGN_DETECTED);
+        }
+
         if (supported_channel_info.dfs_state == beerocks_message::eDfsState::UNAVAILABLE) {
             return AgentDB::sChannelPreference(
                 operating_class, wfa_map::cPreferenceOperatingClasses::ePreference::NON_OPERABLE,
diff --git a/agent/src/beerocks/slave/tasks/channel_selection_task.h b/agent/src/beerocks/slave/tasks/channel_selection_task.h
index d3e45e50f..0f6a3c9d4 100644
--- a/agent/src/beerocks/slave/tasks/channel_selection_task.h
+++ b/agent/src/beerocks/slave/tasks/channel_selection_task.h
@@ -151,6 +151,8 @@ private:
     void handle_vs_channels_list_response(ieee1905_1::CmduMessageRx &cmdu_rx, int fd,
                                           std::shared_ptr<beerocks_header> beerocks_header);
 
+    void handle_vs_awgn_detected_notification(ieee1905_1::CmduMessageRx &cmdu_rx, int fd,
+                                              std::shared_ptr<beerocks_header> beerocks_header);
     void
     handle_vs_zwdfs_ant_channel_switch_response(ieee1905_1::CmduMessageRx &cmdu_rx, int fd,
                                                 std::shared_ptr<beerocks_header> beerocks_header);
@@ -292,6 +294,13 @@ private:
      */
     bool m_send_preference_report_after_csa_finished_event = false;
 
+    /**
+     * AWGN detection event will trigger sending ACTION_BACKHAUL_CHANNELS_LIST_REQUEST.
+     * Set this flag to true to send preference report when handling the response for the
+     * above request in handle_vs_channels_list_response()
+     */
+    bool m_send_preference_report_after_awgn_event = false;
+
     /* Class members */
 
     sPendingChannelSelection m_pending_selection;
diff --git a/common/beerocks/bwl/include/bwl/ap_wlan_hal.h b/common/beerocks/bwl/include/bwl/ap_wlan_hal.h
index edfbdba5c..08b1c1bda 100644
--- a/common/beerocks/bwl/include/bwl/ap_wlan_hal.h
+++ b/common/beerocks/bwl/include/bwl/ap_wlan_hal.h
@@ -69,6 +69,7 @@ public:
         DFS_CAC_Completed,
         DFS_NOP_Finished,
         DFS_RADAR_Detected,
+        AWGN_Detected,
         AP_MGMT_FRAME_RECEIVED,
 
         MGMT_Frame, /**< 802.11 management frame payload */
diff --git a/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h b/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h
index 101da65db..1a81d4a14 100644
--- a/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h
+++ b/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h
@@ -481,6 +481,15 @@ typedef struct {
     sDfsChannelAvailable params;
 } sACTION_APMANAGER_HOSTAP_DFS_CHANNEL_AVAILABLE_NOTIFICATION;
 
+typedef struct {
+    uint32_t frequency;
+    uint8_t channel;
+} sAwgnDetectedParam;
+
+typedef struct {
+    sAwgnDetectedParam params;
+} sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION;
+
 typedef struct {
     int8_t vap_id;
 } sHOSTAP_DISABLED_NOTIFICATION;
diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
index 48f9cf230..886c4d941 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
@@ -113,6 +113,8 @@ static ap_wlan_hal::Event nl80211_to_bwl_event(const std::string &opcode)
 	return ap_wlan_hal::Event::DFS_RADAR_Detected;
     } else if (opcode == "DFS-CAC-START") {
 	return ap_wlan_hal::Event::DFS_CAC_Started;
+    } else if (opcode == "INTERFERENCE-DETECTED") {
+        return ap_wlan_hal::Event::AWGN_Detected;
     } else if (opcode == "CTRL-EVENT-CHANNEL-SWITCH") {
         return ap_wlan_hal::Event::CTRL_Channel_Switch;
     } else if (opcode == "BSS-TM-RESP") {
@@ -3478,6 +3480,35 @@ bool ap_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
 	event_queue_push(event);
 	} break;
 
+    case Event::AWGN_Detected: {
+        // Parse AWGN event: INTERFERENCE-DETECTED type=AWGN
+        std::string interference_type = parsed_obj["type"];
+        if (interference_type != "AWGN") {
+            LOG(DEBUG) << "Ignoring non-AWGN interference type: " << interference_type;
+            break;
+        }
+
+        auto msg_buff =
+            ALLOC_SMART_BUFFER(sizeof(sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION));
+        auto msg = reinterpret_cast<sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION *>(
+                msg_buff.get());
+        LOG_IF(!msg, FATAL) << "Memory allocation failed!";
+
+        // Initialize the message
+        memset(msg_buff.get(), 0, sizeof(sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION));
+
+        // Frequency
+        msg->params.frequency = beerocks::string_utils::stoi(parsed_obj["freq"]);
+
+        // Channel
+        msg->params.channel = son::wireless_utils::freq_to_channel(msg->params.frequency);
+
+        // Add the message to the queue
+        event_queue_push(Event::AWGN_Detected, msg_buff);
+
+        break;
+    }
+
     case Event::Interface_Disabled:
     case Event::ACS_Failed: {
 	 // Forward to the AP manager
diff --git a/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
index 92aa54ab9..44dca3322 100644
--- a/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
@@ -64,6 +64,8 @@ static ap_wlan_hal::Event wpaCtrl_to_bwl_event(const std::string &opcode)
         return ap_wlan_hal::Event::AP_Sta_Possible_Psk_Mismatch;
     } else if (opcode == "ACL-DENY") {
         return ap_wlan_hal::Event::ACL_DENY;
+    } else if (opcode == "INTERFERENCE-DETECTED") {
+        return ap_wlan_hal::Event::AWGN_Detected;
     }
 
     return ap_wlan_hal::Event::Invalid;
@@ -1512,6 +1514,9 @@ bool ap_wlan_hal_whm::process_radio_channel_change_event(const AmbiorixVariant *
         else if (chan_change_reason == "EP_MOVE") {
             LOG(INFO) << "Channel change event received due to backhaul STA endpoint move";
             m_radio_info.last_csa_sw_reason = ChanSwReason::Unknown;
+        } else if ( chan_change_reason == "INITIAL" ) {
+            LOG(INFO) << "Channel change event received due to AWGN";
+            m_radio_info.last_csa_sw_reason = ChanSwReason::Unknown;
         }
         else {
             LOG(ERROR) << "Unsupported channel change reason: " << chan_change_reason;
@@ -2075,6 +2080,34 @@ bool ap_wlan_hal_whm::process_wpa_ctrl_event(const beerocks::wbapi::AmbiorixVari
         event_queue_push(event);
         break;
     }
+    case Event::AWGN_Detected: {
+        // Parse AWGN event: INTERFERENCE-DETECTED type=AWGN
+        std::string interference_type = parsed_obj["type"];
+        if (interference_type != "AWGN") {
+            LOG(DEBUG) << "Ignoring non-AWGN interference type: " << interference_type;
+            break;
+        }
+
+        auto msg_buff =
+            ALLOC_SMART_BUFFER(sizeof(sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION));
+        auto msg = reinterpret_cast<sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION *>(
+                msg_buff.get());
+        LOG_IF(!msg, FATAL) << "Memory allocation failed!";
+
+        // Initialize the message
+        memset(msg_buff.get(), 0, sizeof(sACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION));
+
+        // Frequency
+        msg->params.frequency = beerocks::string_utils::stoi(parsed_obj["freq"]);
+
+        // Channel
+        msg->params.channel = son::wireless_utils::freq_to_channel(msg->params.frequency);
+
+        // Add the message to the queue
+        event_queue_push(Event::AWGN_Detected, msg_buff);
+
+        break;
+    }
     case Event::WPA_Event_EAP_Failure:
     case Event::WPA_Event_EAP_Failure2:
     case Event::WPA_Event_EAP_Timeout_Failure:
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
index 13c496979..5767393ae 100755
--- a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
@@ -154,8 +154,9 @@ eActionOp_BACKHAUL:
   ACTION_BACKHAUL_BSTA_STATE_NOTIFICATION: 87
   ACTION_BACKHAUL_FH_VAPS_STATE_NOTIFICATION: 88
   ACTION_BACKHAUL_SUPPLICANT_RESTART: 89
+  ACTION_BACKHAUL_HOSTAP_AWGN_DETECTED_NOTIFICATION: 90
 
-  ACTION_BACKHAUL_ENUM_END: 90
+  ACTION_BACKHAUL_ENUM_END: 91
 
 ####################################################
 ####################################################
@@ -274,8 +275,9 @@ eActionOp_APMANAGER:
   ACTION_APMANAGER_VAP_DB_UPDATE_NOTIFICATION: 80
   ACTION_APMANAGER_SET_FH_VAPS_STATE: 81
   ACTION_APMANAGER_UPDATE_HIDDEN_SSID_REQUEST: 82
+  ACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION: 83
 
-  ACTION_APMANAGER_ENUM_END: 83
+  ACTION_APMANAGER_ENUM_END: 84
 
 ####################################################
 ####################################################
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml
index 7dc58f490..f14308779 100755
--- a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml
@@ -6,6 +6,7 @@ _include:
     classes/ChannelList.h,
     structs/sCacStartedNotificationParams.h,
     structs/sSmdPolicyConfig.h,
+    structs/sAwgnDetectedParams.h,
     tlvf/WSC/WSC_Attributes.h,
     tlvf/wfa_map/tlvProfile2MultiApProfile.h,
     tlvf/airties/ACSChannelList.h,
@@ -382,3 +383,7 @@ cACTION_APMANAGER_UPDATE_HIDDEN_SSID_REQUEST:
   _type: class
   bssid: sMacAddr
   ssid_advertisement: uint8_t  # 0=SSID hidden (ignore_broadcast_ssid=1), 1=SSID advertised (ignore_broadcast_ssid=0)
+
+cACTION_APMANAGER_HOSTAP_AWGN_DETECTED_NOTIFICATION:
+  _type: class
+  params: sAwgnDetectedParams
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
index 91feadea2..8087aed65 100755
--- a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
@@ -3,6 +3,7 @@
 _include: {
   beerocks/tlvf/beerocks_message_common.h,
   structs/sCacStartedNotificationParams.h,
+  structs/sAwgnDetectedParams.h,
   tlvf/WSC/WSC_Attributes.h,
   tlvf/airties/ACSChannelList.h
 }
@@ -128,6 +129,10 @@ cACTION_BACKHAUL_HOSTAP_DFS_CAC_COMPLETED_NOTIFICATION:
   _type: class
   params: sDfsCacCompleted
 
+cACTION_BACKHAUL_HOSTAP_AWGN_DETECTED_NOTIFICATION:
+  _type: class
+  params: sAwgnDetectedParams
+
 cACTION_BACKHAUL_HOSTAP_ZWDFS_ANT_CHANNEL_SWITCH_REQUEST:
   _type: class
   ant_switch_on: uint8_t #bool
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/structs/sAwgnDetectedParams.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/structs/sAwgnDetectedParams.yaml
new file mode 100644
index 000000000..e703cdb6f
--- /dev/null
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/structs/sAwgnDetectedParams.yaml
@@ -0,0 +1,8 @@
+#
+---
+_namespace: beerocks_message
+
+sAwgnDetectedParams:
+  _type: struct
+  frequency: uint32_t
+  channel: uint8_t
diff --git a/framework/tlvf/yaml/tlvf/wfa_map/tlvChannelPreference.yaml b/framework/tlvf/yaml/tlvf/wfa_map/tlvChannelPreference.yaml
index 5fe2ade85..4da68365b 100644
--- a/framework/tlvf/yaml/tlvf/wfa_map/tlvChannelPreference.yaml
+++ b/framework/tlvf/yaml/tlvf/wfa_map/tlvChannelPreference.yaml
@@ -61,6 +61,7 @@ eReasonCode:
   OPERATION_WOULD_PREVENT_BACKHAUL_OPERATION_USING_SHARED_RADIO: 8
   IMMEDIATE_OPERATION_POSSIBLE_ON_A_DFS_CHANNEL_CAC_HAS_BEEN_RUN__CHANNEL_HAS_BEEN_CLEARED_FOR_USE: 9
   DFS_CHANNEL_STATE_UNKNOWN_CAC_HAS_NOT_RUN: 10
+  AWGN_DETECTED: 14
 
 sFlags:
   _type: struct
-- 
2.34.1

